Plugins: Disable AJAX activation in plugin cards and modals.

costdev opened the pull request:
Testing instructions are found [here in the Trac ticket](https://core.trac.wordpress.org/ticket/61319#comment:2).

Trac ticket:  https://core.trac.wordpress.org/ticket/61319

costdev committed:
Plugins: Disable AJAX activation in plugin cards and modals.

Modals will instead redirect the parent window to the activation URL.

hellofromtonya commented on src/js/_enqueues/wp/updates.js:
```diff
@@ -2664,41 +2664,16 @@
+		 * Click handler for plugin activations in plugin activation modal view.
 		 *
 		 * @since 6.5.0
+		 * @since 6.5.3 Redirect the parent window to the activation URL.
```
```suggestion
		 * @since 6.5.4 Redirect the parent window to the activation URL.
```

costdev committed:
Correct minor version number.

Whoopsie daisy.

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

hellofromtonya reviewed (approved):
Tested with multiple plugins and works as expected ✅ :

- Single plugin with onboarding ✅ 
- Plugin that depends upon a plugin with onboarding ✅ 
- Multiple plugins that depend upon a plugin with onboarding ✅ 
- Single plugin with redirect to a settings or welcome page ✅ 
- Single plugin with no previous redirect ✅ 
- Plugin requiring a plugin with no previous redirect ✅ 

Test Reports in the Trac ticket.

This PR LGTM 👍

afragen commented:
Reverting back to the original AJAX behavior, as this PR accomplishes, is the simplest fix for a minor.

peterwilsoncc commented on src/js/_enqueues/wp/updates.js:
```diff
@@ -2664,41 +2664,16 @@
-		$pluginFilter.on( 'click', '.activate-now', function( event ) {
-			var $activateButton = $( event.target );
-
+		$(document).on( 'click', '#plugin-information-footer .activate-now', function( event ) {
```
I like this as the KISS approach and it tests well.

Rather than use JavaScript to redirect, this entire block of code could be removed/no-opped for the same effect.

costdev replied:
This code is specifically for the modal-based `Activate` buttons which, without this code, will cause navigation _inside_ the modal. instead, this code passes the navigation back to the parent window.

peterwilsoncc replied:
Got it, could you use `target="_parent"` instead or would that break the AJAX-ified update now button?

azaozz replied:
Yea, can probably just remove the whole `$pluginFilter.on( 'click', ...` code block and add `target="_parent"` or `target="_top"` to the link (lines 954 and 998). Did a quick test and it seems to work as expected, but would be better to test it more :)

peterwilsoncc replied:
@costdev and I just chatted about which is the safest approach for the minor. I'm happy with either adding the target parameter in the code below or retaining the JavaScript redirect. Which ever y'all think is safest for inclusion in the minor release.

https://github.com/WordPress/wordpress-develop/blob/19faad774454e86acdd1f0549816cff22b0eb9d8/src/wp-admin/includes/plugin-install.php#L885-L887

I think they are roughly as safe as each other if the button is shown in the modal JS can be assumed to be available, if they target attribute is included the block can be removed.

costdev replied:
Certainly using the native attribute makes sense, though this is also only needed when JS is used as the non-JS version of the modal is its own window (read: `_top` is the same window with non-JS). In that regard, the current implementation in the PR achieves the desired result, only when needed.

What do you think @azaozz? Is it worth making the change? Note this will prompt further testing to verify no additional issues are caused.

azaozz replied:
Yea, the `target` attribute can actually be re-added when the HTML for the button is generated, as it used to be. No need of JS.

> which is the safest approach for the minor

This is the important part. Like @peterwilsoncc I'm thinking either approach is fine. If the current JS redirect has been tested, maybe lets stick with it. Of course more testing would be very welcome, as always, especially for changes in a minor release.

hellofromtonya replied:
tl;dr
I think the safest route at this point is to stay with the JS patch. It's been well tested and has approvals / support.

## Longer version.

If restoring `target` attribute in the Activate HTML `a` and `button` are clean restores, then yes, I agree that is a better way forward. In doing so, it restores the functionality to pre-6.5.0 which was released for a long long while and well-tested.

However, it's not a clean restore:

1. The code to re-add the `target` will be different from pre-6.5 (see below).
2. Prior to 6.5, the `target` was `_parent`. There's discussion above about possibly changing that to `_top`.

These changes will need consideration and re-testing.

### Code is different.

Prior to 6.5, the `target` was part of the `a` and `button` HTML echoed to the browser. 6.5.0 changed it, where the HTML is built in a new function `wp_get_plugin_action_button()` and then `str_replace()` modifies the HTML before passing it into `wp_kses_post()` and echoing it out to the browser.

6.5:
https://github.com/WordPress/wordpress-develop/blob/19faad774454e86acdd1f0549816cff22b0eb9d8/src/wp-admin/includes/plugin-install.php#L885-L887

6.4:
https://github.com/WordPress/wordpress-develop/blob/db589b327cf551071cfee14a6be36e8552321db9/src/wp-admin/includes/plugin-install.php#L891-L893

Where should the `target` be re-added? if within the `str_contains( ... 'Activate' ... )` block, does it need to check first if a `target` exists before adding or changing it?

### My point
It doesn't appear to be a straight re-adding of the `target`, but rather new, untested code that needs re-review and testing.

I'm thinking the JS approach is the way to go for the minor due to:
* is already tested across many different plugins (though retesting of these plugins could be done)
* only targets the modal's `Activate` button
* today is the scheduled RC1 day (though that could be changed)

mukeshpanchal27 commented on src/js/_enqueues/wp/updates.js:
```diff
@@ -2664,41 +2664,16 @@
-		$pluginFilter.on( 'click', '.activate-now', function( event ) {
-			var $activateButton = $( event.target );
-
+		$(document).on( 'click', '#plugin-information-footer .activate-now', function( event ) {
```
```suggestion
		$document.on( 'click', '#plugin-information-footer .activate-now', function( event ) {
```

mukeshpanchal27 replied:
@costdev small suggestion.

hellofromtonya replied:
That does fit into the rest of the file for consistency. Good catch @mukeshpanchal27 👍

hellofromtonya replied:
Tested - just to make sure. Yes, works with no change in behavior 👍

hellofromtonya committed:
Use the var for document for consistency.

The `var $document` is `$( document )`. The variable is used throughout the same script file. Changing for consistency.

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/58250. Thank you everyone for contributing 🌟
